src/pages/Login.jsx

```diff
@@ -1,21 +1,117 @@
-import React from "react";
+import React, { useState, useEffect } from "react";
+import { useStateCallback } from "use-state-callback";
 import "../styles/Authentication.css";
+import { useNavigate, useLocation } from "react-router-dom";
+import { validateLoginUser, loginUser, useUser } from "../state/index";
+import { toast } from "react-toastify";
+import { USER_AUTH_ACTION } from "../state/action";
+
 const Login = () => {
+  const [password, setPassword] = useStateCallback("");
+  const [email, setEmail] = useStateCallback("");
+  const [emailError, setEmailError] = useState("");
+  const [passwordError, setPasswordError] = useStateCallback("");
+  const [showPassword, setShowPassword] = useState(false);
+  const navigate = useNavigate();
+  const { state } = useLocation();
+
+  const redirect = state?.from || "/";
+
+  const {
+    user: { isLoggedIn },
+    userDispatch,
+  } = useUser();
+
+  const loginAndValidateUser = async () => {
+    const validate = validateLoginUser(email, emailError, passwordError);
+    if (!validate) {
+      toast.error("Please fill all the details");
+      return;
+    }
+
+    const res = await loginUser(email, password);
+    const { status, foundUser, encodedToken } = res;
+
+    if (status === 201) {
+      toast.error("wrong password");
+      return;
+    }
+    if (status === 404) {
+      toast.error("wrong email and password");
+      return;
+    }
+    toast.success(`${foundUser.firstName} wish you happy shopping`);
+    const loginDetails = {
+      isLoggedIn: true,
+      firstName: foundUser.firstName,
+      encodedToken: encodedToken,
+    };
+
+    localStorage.setItem("loginDetails", JSON.stringify(loginDetails));
+
+    userDispatch({
+      type: USER_AUTH_ACTION.LOGIN_SUCCESSFULL,
+      payload: { ...loginDetails },
+    });
+
+    navigate(redirect, { replace: true });
+  };
+  useEffect(
+    () => isLoggedIn && navigate(redirect, { replace: true }),
+    [isLoggedIn]
+  );
   return (
     <main className="signup-container">
       <section className="signup-box">
         <h1 className="signup-header">Sign In</h1>
 
         <div className="lab-input">
           <label>Email*</label>
-          <input type="Email" placeholder="shashimourya1@gmail.com" />
+          <input
+            type="Email"
+            placeholder="shashimourya1@gmail.com"
+            value={email}
+            onChange={(e) =>
+              setEmail(e.target.value, (email) => {
+                const reg = /\S+@\S+\.\S+/;
```

Comment:
Instead of setting it here add it inside a function handler, it will look much cleaner.

Reply:
instead of useState i am using useStateCallback i found this easy to maintain .thanks for suggestion i will look into it:)